pinot-common/src/test/java/org/apache/pinot/common/metrics/AbstractMetricsTest.java

```diff
@@ -108,4 +112,154 @@ public void testMultipleGauges() {
+   */
+  @Test
+  public void testMeteredMetrics() {
+    final ControllerMetrics testMetrics = buildTestMetrics();
+    final MetricsInspector inspector = new MetricsInspector(testMetrics.getMetricsRegistry());
+    final String tableName = "tbl_testMeteredMetrics";
+    final String keyName = "keyName";
+    final ControllerMeter meter = ControllerMeter.CONTROLLER_INSTANCE_POST_ERROR;
+    final ControllerMeter meter2 = ControllerMeter.CONTROLLER_PERIODIC_TASK_ERROR;
+
+    // Holder for the most recently seen Metric
+    final MetricName[] currentMetric = new MetricName[1];
+
+    // When a new metric is expected we'll call this lambda to assert the metric was created, and update currentMetric
+    final Runnable expectNewMetric = () -> {
+      Assert.assertNotEquals(inspector.lastMetric(), currentMetric[0]);
+      currentMetric[0] = inspector.lastMetric();
+    };
+
+    // Lambda to verify that the latest metric has the expected value, and its creation was expected
+    final IntConsumer expectMeteredCount = expected -> {
+      Assert.assertEquals(inspector.getMetered(currentMetric[0]).count(), expected);
+      Assert.assertEquals(currentMetric[0], inspector.lastMetric());
+    };
+
+    // Test global meter APIs
+    testMetrics.addMeteredGlobalValue(meter, 5);
+    expectNewMetric.run();
+    expectMeteredCount.accept(5);
+    testMetrics.addMeteredGlobalValue(meter, 4, testMetrics.getMeteredValue(meter));
+    expectMeteredCount.accept(9);
+
+    // Test meter with key APIs
+    testMetrics.addMeteredValue(keyName, meter, 9);
+    expectNewMetric.run();
+    PinotMeter reusedMeter = testMetrics.addMeteredValue(keyName, meter2, 13, null);
+    expectNewMetric.run();
+    expectMeteredCount.accept(13);
+    testMetrics.addMeteredValue(keyName, meter2, 6, reusedMeter);
+    expectMeteredCount.accept(19);
+
+    // Test table-level meter APIs
+    testMetrics.addMeteredTableValue(tableName, meter, 15);
+    expectNewMetric.run();
+    expectMeteredCount.accept(15);
+    testMetrics.addMeteredTableValue(tableName, meter2, 3, testMetrics.getMeteredTableValue(tableName, meter));
+    expectMeteredCount.accept(18);
+
+    // Test table-level meter with additional key APIs
+    testMetrics.addMeteredTableValue(tableName, keyName, meter, 21);
+    expectNewMetric.run();
+    expectMeteredCount.accept(21);
+    reusedMeter = testMetrics.addMeteredTableValue(tableName, keyName, meter2, 23, null);
+    expectNewMetric.run();
+    expectMeteredCount.accept(23);
+    testMetrics.addMeteredTableValue(tableName, keyName, meter2, 5, reusedMeter);
+    expectMeteredCount.accept(28);
+
+    // Test removal APIs
+    Assert.assertEquals(testMetrics.getMetricsRegistry().allMetrics().size(), 6);
+    // This is the only AbstractMetrics method for removing Meter-type metrics. Should others be added?
+    testMetrics.removeTableMeter(tableName, meter);
+    Assert.assertEquals(testMetrics.getMetricsRegistry().allMetrics().size(), 5);
```

Comment:
let's clean all the meters and check if the metrics registry is empty here as well

Reply:
Done.